### PR TITLE
device/trezor: integration tests

### PR DESCRIPTION
**Pull request by ph4r05**
This PR adds integration tests with the Trezor emulator. 

- Tests Key image sync and transaction signing.  
- Several transaction signing scenarios with varying #utxo, types of utxo (received to sub-address, received with additional tx keys), # of outputs, types of outputs (normal, to sub-address, to multiple addresses), payment id.

## How it works
- Tests generate valid blockchain (init step) and then spends mined funds with the Trezor (test step). 
  - *Init step*: In order to save CPU time, the common blockchain is generated before all tests are run. Before each test, the blockchain is rewound to the original state. Core verifies only new blocks on the blockchain added by the test being run. 
  - *Test step*: There are several testing scenarios added for testing transaction signature on Trezor. Scenarios differ in number of UTXO, outputs and other transaction parameters to verify multiple TX signing codepaths. Testing Trezor transactions are generated according to the test parameters by a TrensactionBuilder and sent to Trezor for signing.
  - Each transaction test tries to spend some funds by signing TX with the Trezor. Transaction is later verified.
- Generated transactions are validated by the Monero core (daemon logic), like tests that are already in place.
- Generated transactions are checked whether they are well-formed and whether the received funds are spendable. Tests directly use `tools::wallet2` for this.

I needed to get RCT and Bulletproof transaction working with `chaingen.cpp` functions so I had to do some overhaul of the `chaingen`. The changes can either stay as they are or I can move this modified version of `chaingen` to the trezor tests directory. Depends on the reviews.

The common blockchain created at the beginning of the Trezor tests can be saved to a file so the it is not generated with each test run. The argument `--chain_path` provides path where to save/load the common chain. The #5122 is required for chain save/load work correctly.

## Testing

You first need to get Trezor emulator running. Ideally follow official [building documentation](https://github.com/trezor/trezor-core/blob/master/docs/build.md).

Emulator can run in the Docker. I will add instructions later.

Summary for Trezor emulator build:

```
git clone --recursive https://github.com/trezor/trezor-core.git
cd trezor-core
make vendor

# ubuntu / debian:
sudo apt-get install scons libsdl2-dev libsdl2-image-dev

# fedora:
sudo yum install scons SDL2-devel SDL2_image-devel

# osx:
brew install scons sdl2 sdl2_image pkg-config

# build and start emulator
make build_unix

# for build without UI run
make build_unix_noui

# run emulator
PYOPT=0 make emu
```

To build and run Monero integration tests:

```
cd -
make debug-test-trezor
./build/$os/$branch/debug/tests/trezor/trezor_tests
```

### CODE REVIEW

**Comment by ph4r05** [comment]
After going through the PR again I am seriously thinking about small refactoring, namely:

- Separate changes to separate chaingen files, just for trezor tests. So it is independent, added as a separate file. This will be much easier to review and less prone to errors and regressions as now - large amount of edits in existing chaingen.

- Wallet2 friend class thining. I need to use Wallet in the tests and to call small number of private methods. Now I use testing friend class for that. I am thinking about reducing this to minimum with clean API - private methods will be re-exported in the `Wallet_Accessor_Test`, while the test wallet logic will be moved to a separate class so accessor does really only testing adapter on private methods.

What do you think @moneromooo-monero ?

File: src/device_trezor/device_trezor_base.hpp [hunk 1]
```diff
@@ -229,6 +229,12 @@ namespace trezor {
       return m_features;
     }
 
+    uint64_t get_version() {
+      CHECK_AND_ASSERT_THROW_MES(m_features, "Features not loaded");
+      CHECK_AND_ASSERT_THROW_MES(m_features->has_major_version() && m_features->has_minor_version() && m_features->has_patch_version(), "Invalid Trezor firmware version information");
+      return pack_version(m_features->major_version(), m_features->minor_version(), m_features->patch_version());
+    }
```
Comment (moneromooo-monero) [inline src/device_trezor/device_trezor_base.hpp]: const ?

File: src/device_trezor/trezor/transport.hpp [hunk 2]
```diff
@@ -62,6 +62,8 @@ namespace trezor {
 
   const std::string DEFAULT_BRIDGE = "127.0.0.1:21325";
 
+  uint64_t pack_version(uint32_t major, uint32_t minor=0, uint32_t patch=0);
+
```
Comment (moneromooo-monero) [inline src/device_trezor/trezor/transport.hpp]: const ?

File: src/device_trezor/trezor/transport.cpp [hunk 3]
```diff
@@ -660,8 +685,7 @@ namespace trezor{
   WebUsbTransport::WebUsbTransport(
       boost::optional<libusb_device_descriptor*> descriptor,
       boost::optional<std::shared_ptr<Protocol>> proto
-  ): m_conn_count(0),
-     m_usb_session(nullptr), m_usb_device(nullptr), m_usb_device_handle(nullptr),
+  ): m_usb_session(nullptr), m_usb_device(nullptr), m_usb_device_handle(nullptr),
```
Comment (moneromooo-monero) [inline src/device_trezor/trezor/transport.cpp]: m_open_counter does not get initiliazed now ?

File: src/device_trezor/device_trezor.cpp [hunk 4]
```diff
@@ -97,7 +97,7 @@ namespace trezor {
         auto res = get_view_key();
         CHECK_AND_ASSERT_MES(res->watch_key().size() == 32, false, "Trezor returned invalid view key");
 
-        spendkey = crypto::null_skey; // not given
+        //spendkey = crypto::null_skey; // not given
```
Comment (moneromooo-monero) [inline src/device_trezor/device_trezor.cpp]: Intended ? If it's uninitialized, maybe it should not be a parameter at all in the first place.

File: src/device_trezor/device_trezor_base.cpp [hunk 5]
```diff
@@ -355,6 +368,32 @@ namespace trezor {
...
+  if (m_debug_callback) m_debug_callback->method(__VA_ARGS__); \
+  if (m_callback) m_callback->method(__VA_ARGS__);             \
+}while(0)
+
+    void device_trezor_base::setup_debug(){
+      if (!m_debug){
+        return;
+      }
+
+      if (m_debug && !m_debug_callback){
```
Comment (moneromooo-monero) [inline src/device_trezor/device_trezor_base.cpp]: That was tested just above

File: src/device_trezor/device_trezor_base.cpp [hunk 6]
```diff
@@ -355,6 +368,32 @@ namespace trezor {
...
+        if (debug_transport) {
+          m_debug_callback = std::make_shared<trezor_debug_callback>(debug_transport);
+        } else {
+          MDEBUG("Transport does not have debug link option");
+        }
+      }
+    }
+
+#else
+#define TREZOR_CALLBACK(method, ...) if (m_callback) m_callback->method(__VA_ARGS__)
```
Comment (moneromooo-monero) [inline src/device_trezor/device_trezor_base.cpp]: That could do with a do while(0) too.

File: src/device_trezor/device_trezor_base.cpp [hunk 7]
```diff
@@ -421,5 +453,67 @@ namespace trezor {
...
+      init_device();
+    }
+
+    void device_trezor_base::init_device()
+    {
+      auto msg = std::make_shared<messages::management::Initialize>();
+      m_features = client_exchange<messages::management::Features>(msg);
+    }
+
+    void device_trezor_base::load_device(const std::string & mnemonic, const std::string & pin,
```
Comment (moneromooo-monero) [inline src/device_trezor/device_trezor_base.cpp]: wipeable_string ?

File: src/device_trezor/trezor/transport.cpp [hunk 8]
```diff
@@ -223,6 +223,30 @@ namespace trezor{
...
+  void Transport::session_end(){
+    m_session_counter -= 1;
+
+    if (m_session_counter < 0){
+      MDEBUG("Session already closed. ctr: " << m_session_counter);
+
+    } else if (m_session_counter == 0) {
+      close();
+    }
+  }
```
Comment (moneromooo-monero) [inline src/device_trezor/trezor/transport.cpp]: Why am I feeling that there's some amount of duplicate logic in all this :/

File: src/wallet/wallet2.h [hunk 9]
```diff
@@ -67,6 +67,7 @@
 #define MONERO_DEFAULT_LOG_CATEGORY "wallet.wallet2"
 
 class Serialization_portability_wallet_Test;
+class Wallet_Accessor_Test;
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.h]: Caps and underscores. That's some new naming scheme. The above class isn't used anywhere fwiw, it can be removed.

**Comment by ph4r05** [comment]
Thanks for the review. I am a bit slammed at the moment but I will get to that on 28th 1.

**Comment by moneromooo-monero** [comment]
Tests: this is pretty chunky, but it seems OK overall, though I read through quickly. Feel free to do whatever cleanup you mentioned, but overall this seems fine. Some explanations of the main changes to the core test machinery would be nice, the core test code is hard to understand (for me anyway).

File: src/device_trezor/device_trezor_base.cpp (hunk 7 above)
Comment (ph4r05) [reply src/device_trezor/device_trezor_base.cpp]: This function is applicable only with debugging Trezor firmware, basically with an emulator for integration tests. It cannot be used on production so I thought normal strings would do the job here.

File: src/device_trezor/trezor/transport.cpp (hunk 3 above)
Comment (ph4r05) [reply src/device_trezor/trezor/transport.cpp]: It was removed from this obj hierarchy

File: src/device_trezor/trezor/transport.hpp (hunk 2 above)
Comment (ph4r05) [reply src/device_trezor/trezor/transport.hpp]: it is a plain function in the namespace. so OK as it is?

File: src/device_trezor/device_trezor_base.hpp (hunk 1 above)
Comment (ph4r05) [reply src/device_trezor/device_trezor_base.hpp]: it is a plain function in the namespace. so OK as it is?

File: src/device_trezor/trezor/transport.cpp (hunk 8 above)
Comment (ph4r05) [reply src/device_trezor/trezor/transport.cpp]: I've fixed that by moving the open/close counting to `pre_open` and `pre_close` methods.

**Comment by ph4r05** [comment]
I am working on getting tests back running after the rebase. I will ping once it's done.

**Comment by ph4r05** [comment]
OK tests are passing again

**Comment by ph4r05** [comment]
I've added chain serialization/deserialization so the initial blockchain that is required for subsequent payment tests is not generated with each test run. The command `--chain_path` provides path where to save/load the common chain.

The https://github.com/monero-project/monero/pull/5122 is required for chain save/load work correctly.

**Comment by moneromooo-monero** [comment]
> I needed to get RCT and Bulletproof transaction working with chaingen.cpp functions so I had to do some overhaul of the chaingen.

That seems odd. There are rct and BP core tests, and they pass.

**Comment by ph4r05** [comment]
> > I needed to get RCT and Bulletproof transaction working with chaingen.cpp functions so I had to do some overhaul of the chaingen.
> 
> That seems odd. There are rct and BP core tests, and they pass.

As far as I know those tests are implemented separately. `Chaingen.{h,cpp}` did not provide API to generate RCT transactions on the blockchain. Like there is a macro `MAKE_TX_MIX_LIST` which creates a transaction that is later added to the block. I made `MAKE_TX_MIX_LIST_RCT` which enables to generate v2 transactions on the blockchain, with bulletproofs. The idea of this is to spend rctv2 transactions in Trezor, which I could not generate with changen before (alternative was to do it manually).

**Comment by moneromooo-monero** [comment]
Ah, OK, that might be true.
I'll review that when I have some time for it, hopefully soon.

File: tests/core_tests/chaingen.cpp [hunk 10]
```diff
@@ -477,45 +450,627 @@ bool fill_tx_sources(std::vector<tx_source_entry>& sources, const std::vector<te
...
+
+  for(size_t j=0; j < tx->vout.size(); ++j){
+    auto it = find_out(tx->hash, j);
+    if (it != m_map_outs.end()){
+      indices.push_back(it->second.idx);
+    }
+  }
+}
+
+void block_tracker::get_decoys(size_t decoys, uint64_t amount, uint64_t global_index, uint64_t cur_height, std::vector<tools::wallet2::get_outs_entry> &outs){
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply tests/core_tests/chaingen.cpp]: You're right, makes sense. Fixed.

File: tests/core_tests/chaingen.cpp [hunk 11]
```diff
@@ -477,45 +450,627 @@ bool fill_tx_sources(std::vector<tx_source_entry>& sources, const std::vector<te
...
+    idx = (idx + 1) % n_outs;
+    size_t oi_idx = choices[(size_t)idx];
+    CHECK_AND_ASSERT_THROW_MES((n_iters / n_outs) <= outs.size(), "Decoy pick selection problem");
+
+    auto & oi = vct[oi_idx];
+    if (oi.idx == global_index)
+      continue;
+    if (oi.out.type() != typeid(cryptonote::txout_to_key))
+      continue;
+    if (oi.blk_height + CRYPTONOTE_MINED_MONEY_UNLOCK_WINDOW > cur_height)
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply tests/core_tests/chaingen.cpp]: True, I wanted to have a safe choice for fake outputs.

[the remark this answers is not in the document]
Comment (ph4r05) [reply tests/core_tests/chaingen.cpp]: I've fixed it by storing unlock_height to `oi` so the calculation is precise now.

**Comment by ph4r05** [comment]
Wallet2 is quite used in the `chaingen` now, i.e., creating transactions by passing wallet as a source of the transaction. Would you rather prefer to separate wallet-related code to a new file?

**Comment by moneromooo-monero** [comment]
I think it makes more sense, yes. Thanks.

**Comment by ph4r05** [comment]
I like it more like this also. Squashed.

**Comment by moneromooo-monero** [comment]
Anyone else planning to review this ?

File: tests/core_tests/wallet_tools.cpp [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+#include <random>
+
+using namespace std;
+using namespace epee;
+using namespace crypto;
+using namespace cryptonote;
+
+// Shared random generator
+static std::random_device RD;
+static std::mt19937 MT(RD());
```
Comment (mudlotus) [inline tests/core_tests/wallet_tools.cpp]: Know this is test code, but this isn't a CSRNG. What do you think about using a [libsodium custom-rng](https://download.libsodium.org/doc/advanced/custom_rng), or Monero's RNG?

Might prevent accidental use of Mersenne-twister in production code.

Comment (ph4r05) [reply tests/core_tests/wallet_tools.cpp]: OK for peace of mind I've changed it to use `crypto::rand` and for shuffling `std::default_random_engine(crypto::rand<unsigned>())`, the same as in the production code (e.g., tx output shuffling).

**Comment by ph4r05** [comment]
@moneromooo-monero I noticed a discussion on the IRC and I think these tests could be easily modified to support also Ledger. The core could be the same, just the transaction signature would go to the Ledger instead of the Trezor.

**Comment by moneromooo-monero** [comment]
Are you saying you're willing to do that ? If so that's great :) This is probably going to get merged soon, since I kinda doubt anyone is really going to review as well, so another PR on top would be best, unless you think a lot of shuffling around is needed first.

**Comment by ph4r05** [comment]
I can certainly help with that but I am busy to do it on my own 😅 
It was more like suggestion for Ledger guys. 

Regarding the shuffling - I think this is OK for now. The heavy changes has been made in `chaingen` and `wallet_tools`.

**Pull request merged**